File: cuhara.qua.back/src/modules/users/commands/create-user.ts
import { BaseCommand, CommandHandler } from "../../../core/cqrs/command";
import { PrismaClient } from "@prisma/client";
import { prisma } from "../../../prisma/client";

// ==================== REQUEST ====================
export interface CreateUserRequest {
  name: string;
  email: string;
  vscAccount: string;
  roleId: number;
  tenantId: number;
}

// ==================== RESPONSE ====================
export interface CreateUserResponse {
  id: number;
}

// ==================== COMMAND ====================
export class CreateUserCommand extends BaseCommand<CreateUserResponse> {
  readonly type = "CreateUserCommand";

  constructor(public readonly data: CreateUserRequest) {
    super();
  }
}

// ==================== HANDLER ====================
export class CreateUserHandler
  implements CommandHandler<CreateUserCommand, CreateUserResponse>
{
  private prisma: PrismaClient;

  constructor() {
    this.prisma = new PrismaClient();
  }

  async handle(command: CreateUserCommand): Promise<CreateUserResponse> {
    try {
      const user = await this.prisma.user.create({
        data: {
          name: command.data.name,
          email: command.data.email,
          vscAccount: command.data.vscAccount,
          roleId: command.data.roleId,
          tenantId: command.data.tenantId,
        },
      });

      return {
        id: user.id,
      };
    } catch (error) {
      throw new Error(`Failed to create user: ${error}`);
    }
  }
}

File: cuhara.qua.go/internal/users/application/command/create_user.go
package command

import (
	"context"

	"cuhara.qua.go/internal/common/cqrs"
	"cuhara.qua.go/internal/users/domain"
)

const CreateUserCommandType = "user.create"

type CreateUserResponse struct {
	ID int64 `json:"id" example:"1"`
}

type CreateUserRequest struct {
	Name       string `json:"name" example:"John Doe" binding:"required"`
	Email      string `json:"email" example:"john@example.com" binding:"required,email"`
	VscAccount string `json:"vsc_account" example:"john_doe"`
	RoleID     int64  `json:"role_id" example:"1" binding:"required"`
	TenantID   int64  `json:"tenant_id" example:"1" binding:"required"`
}

type CreateUserCommand struct {
	cqrs.BaseCommand[CreateUserResponse]

	Name       string
	Email      string
	VscAccount string
	RoleID     int64
	TenantID   int64
}

func NewCreateUserCommand(r CreateUserRequest) *CreateUserCommand {
	command := cqrs.NewBaseCommand[CreateUserResponse](CreateUserCommandType)
	return &CreateUserCommand{
		BaseCommand: command,
		Name:        r.Name,
		Email:       r.Email,
		VscAccount:  r.VscAccount,
		RoleID:      r.RoleID,
		TenantID:    r.TenantID,
	}
}

type CreateUserHandler struct {
	repo domain.UserRepository
}

func NewCreateUserHandler(repo domain.UserRepository) *CreateUserHandler {
	return &CreateUserHandler{repo: repo}
}

func (h *CreateUserHandler) Handle(ctx context.Context, c cqrs.Command[CreateUserResponse]) (CreateUserResponse, error) {
	cmd := c.(*CreateUserCommand)

	u, events, err := domain.NewUser(cmd.Name, cmd.Email, cmd.VscAccount, cmd.RoleID, cmd.TenantID)
	if err != nil {
		return CreateUserResponse{}, err
	}

	id, err := h.repo.Create(ctx, u, events)
	if err != nil {
		return CreateUserResponse{}, err
	}

	return CreateUserResponse{ID: id}, nil
}
